# Notebook para los códigos del TFG

## Librerias necesarias

### Instalamos las librerias

In [2]:
#%pip install pandas
#%pip install pysr

In [3]:
%lsmagic

Available line magics:
%alias  %alias_magic  %autoawait  %autocall  %automagic  %autosave  %bookmark  %cd  %clear  %cls  %code_wrap  %colors  %conda  %config  %connect_info  %copy  %ddir  %debug  %dhist  %dirs  %doctest_mode  %echo  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %macro  %magic  %mamba  %matplotlib  %micromamba  %mkdir  %more  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %pip  %popd  %pprint  %precision  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %ren  %rep  %rerun  %reset  %reset_selective  %rmdir  %run  %save  %sc  %set_env  %store  %sx  %system  %tb  %time  %timeit  %unalias  %unload_ext  %uv  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%capture  %%cmd  %%code_wrap  %%debug  %%file  %%html  %%javascript  %%js  %%l

### Importamos las librerias

In [4]:
import pandas as pd
#from pysr import PySRRegressor De momento no lo estoy usando
import seaborn as sns
import matplotlib.pyplot as plt

## Preprocesamiento de datos

### Abrir los archivos

#### Abrimos el archivo physical properties

In [5]:
physical_properties = 'Data/raw/hlsp_candels_hst_wfc3_egs_multi_v1_physpar-cat.txt'

phys_prop_data = pd.read_csv(physical_properties, sep='\s+', header=None, skiprows=123)



#A continuación añadimos las columnas

phys_prop_data.columns = ["ID",
"age_2a_tau (dex(yr))" ,
"tau_2a_tau (Gyr)" ,
"Av_2a_tau (mag)" ,
"SFR_2a_tau (solMass/yr)" ,
"chi2_2a_tau ",
"age_4b (dex(yr))" ,
"EBV_4b (mag)" ,
"age_6a_tau (dex(yr))" ,
"tau_6a_tau (Gyr)" ,
"EBV_6a_tau (mag)" ,
"SFR_6a_tau (solMass/yr)" ,
"met_6a_tau (solMet)" ,
"extlw_6a_tau ",
"chi2_6a_tau ",
"L1400_6a_tau (erg/s/Hz)" ,
"L2700_6a_tau (erg/s/Hz)" ,
"UMag_6a_tau (mag)" ,
"BMag_6a_tau (mag)" ,
"VMag_6a_tau (mag)" ,
"RMag_6a_tau (mag)" ,
"IMag_6a_tau (mag)" ,
"JMag_6a_tau (mag)" ,
"KMag_6a_tau (mag)" ,
"age_10c (dex(yr))" ,
"SFH_10c ",
"tau_10c (Gyr)" ,
"met_10c (solMet)" ,
"M_l99_11a_tau (dex(solMass))",
 "M_u99_11a_tau (dex(solMass))",
 "age_11a_tau (dex(yr))" ,
"SFR_11a_tau (solMass/yr)" ,
"M_l68_12a (dex(solMass))",
"M_u68_12a (dex(solMass))",
 "M_l95_12a (dex(solMass))",
 "M_u95_12a (dex(solMass))",
 "age_12a (dex(yr))" ,
"tau_12a (Gyr)" ,
"EBV_12a (mag)" ,
"met_12a (solMet)" ,
"Lbol_12a (dex(solLum))" ,
"chi2_12a ",
"age_13a_tau (dex(yr))" ,
"tau_13a_tau (Gyr)" ,
"Av_13a_tau (mag)" ,
"SFR_13a_tau (solMass/yr)" ,
"chi2_13a_tau ",
"age_14a (dex(yr))" ,
"SFH_14a ",
"tau_14a (Gyr)" ,
"EBV_14a (mag)" ,
"SFR_14a (solMass/yr)" ,
"q_14a ",
"age_6a_tau ^NEB (dex(yr))" ,
"tau_6a_tau ^NEB (Gyr)" ,
"EBV_6a_tau ^NEB (mag)" ,
"SFR_6a_tau ^NEB (solMass/yr)" ,
"met_6a_tau ^NEB (solMet)" ,
"extlw_6a_tau ^NEB ",
"chi2_6a_tau ^NEB ",
"L1400_6a_tau ^NEB (erg/s/Hz)" ,
"L2700_6a_tau ^NEB (erg/s/Hz)" ,
"UMag_6a_tau ^NEB (mag)" ,
"BMag_6a_tau ^NEB (mag)" ,
"VMag_6a_tau ^NEB (mag)" ,
"RMag_6a_tau ^NEB (mag)" ,
"IMag_6a_tau ^NEB (mag)" ,
"JMag_6a_tau ^NEB (mag)" ,
"KMag_6a_tau ^NEB (mag)" ,
"age_6a_deltau (dex(yr))" ,
"tau_6a_deltau (Gyr)" ,
"EBV_6a_deltau (mag)" ,
"SFR_6a_deltau (solMass/yr)" ,
"met_6a_deltau (solMet)" ,
"extlw_6a_deltau ",
"chi2_6a_deltau ",
"L1400_6a_deltau (erg/s/Hz)" ,
"L2700_6a_deltau (erg/s/Hz)" ,
"UMag_6a_deltau (mag)" ,
"BMag_6a_deltau (mag)" ,
"VMag_6a_deltau (mag)" ,
"RMag_6a_deltau (mag)" ,
"IMag_6a_deltau (mag)" ,
"JMag_6a_deltau (mag)" ,
"KMag_6a_deltau (mag)" ,
"age_6a_invtau (dex(yr))" ,
"tau_6a_invtau (Gyr)" ,
"EBV_6a_invtau (mag)" ,
"SFR_6a_invtau (solMass/yr)" ,
"met_6a_invtau (solMet)" ,
"extlw_6a_invtau ",
"chi2_6a_invtau ",
"L1400_6a_invtau (erg/s/Hz)" ,
"L2700_6a_invtau (erg/s/Hz)" ,
"UMag_6a_invtau (mag)" ,
"BMag_6a_invtau (mag)" ,
"VMag_6a_invtau (mag)" ,
"RMag_6a_invtau (mag)" ,
"IMag_6a_invtau (mag)" ,
"JMag_6a_invtau (mag)" ,
"KMag_6a_invtau (mag)" ,
"age_10c ^dust (dex(yr))" ,
"SFH_10c ^dust ",
"tau_10c ^dust (Gyr)" ,
"met_10c ^dust (solMet)" ,
"age_14a_const (dex(yr))" ,
"EBV_14a_const (mag)" ,
"SFR_14a_const (solMass/yr)" ,
"q_14a_const ",
"age_14a_lin (dex(yr))" ,
"EBV_14a_lin (mag)" ,
"SFR_14a_lin (solMass/yr)" ,
"q_14a_lin ",
"age_14a_deltau (dex(yr))" ,
"tau_14a_deltau (Gyr)" ,
"EBV_14a_deltau (mag)" ,
"SFR_14a_deltau (solMass/yr)" ,
"q_14a_deltau ",
"age_14a_tau (dex(yr))" ,
"tau_14a_tau (Gyr)" ,
"EBV_14a_tau (mag)" ,
"SFR_14a_tau (solMass/yr)" ,
"q_14a_tau "]

##### Lo mostramos

In [6]:
phys_prop_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41457 entries, 0 to 41456
Columns: 123 entries, ID to q_14a_tau 
dtypes: float64(110), int64(13)
memory usage: 38.9 MB


#### Abrimos el archivo mass

In [7]:
mass = 'Data/raw/hlsp_candels_hst_wfc3_egs_multi_v1_mass-cat.txt'

mass_data = pd.read_csv(mass, sep='\s+', header=None, skiprows=39)


#A continuación añadimos las columnas

mass_data.columns = ["ID"
,"RAdeg (deg) "
,"DECdeg (deg) "
,"Hmag (mag) "
,"PhotFlag "
,"Class_star "
,"AGNflag "
,"zphot "
,"zspec "
,"q_zspec "
,"r_zspec "
,"zbestM "
,"zphot_l68 "
,"zphot_u68 "
,"zphot_l95 "
,"zphot_u95 "
,"zAGN "
,"M_neb_med (dex(solMass)) "
,"s_neb_med (dex(solMass)) "
,"M_med (dex(solMass)) "
,"s_med (dex(solMass)) "
,"M_14a_cons (dex(solMass)) "
,"M_11a_tau (dex(solMass)) "
,"M_6a_tau^NEB (dex(solMass)) "
,"M_13a_tau (dex(solMass)) "
,"M_12a_tau (dex(solMass)) "
,"M_6a_tau (dex(solMass)) "
,"M_2a_tau (dex(solMass)) "
,"M_15a (dex(solMass)) "
,"M_10c (dex(solMass)) "
,"M_14a_lin (dex(solMass)) "
,"M_14a_deltau (dex(solMass)) "
,"M_14a_tau (dex(solMass)) "
,"M_14a_inctau (dex(solMass)) "
,"M_14a (dex(solMass)) "
,"M_neb_med_lin (solMass) "
,"s_neb_med_lin (solMass) "
,"M_med_lin (solMass) "
,"s_med_lin (solMass)" ]

##### Lo mostramos

In [8]:
print(mass_data['AGNflag '].value_counts())
print("\n")
print(sum(list(mass_data['zAGN '].value_counts()))-41215)

AGNflag 
0    41215
1      242
Name: count, dtype: int64


242


#### Abrimos el archivo photo

In [9]:
phot = 'Data/raw/hlsp_candels_hst_wfc3_egs_multi_v2_redshift-cat.txt'

phot_data = pd.read_csv(phot, sep='\s+', header=None, skiprows=61)


#A continuación añadimos las columnas

phot_data.columns = [" file "
,"ID"
,"RA"
,"DEC"
,"z_best "
,"z_best_type "
,"z_spec "
,"z_spec_ref "
,"z_grism "
,"mFDa4_z_peak "
,"mFDa4_z_weight "
,"mFDa4_z683_low "
,"mFDa4_z683_high "
,"mFDa4_z954_low "
,"mFDa4_z954_high "
,"HB4_z_peak "
,"HB4_z_weight "
,"HB4_z683_low "
,"HB4_z683_high "
,"HB4_z954_low "
,"HB4_z954_high "
,"Finkelstein_z_peak "
,"Finkelstein_z_weight "
,"Finkelstein_z683_low "
,"Finkelstein_z683_high "
,"Finkelstein_z954_low "
,"Finkelstein_z954_high "
,"Fontana_z_peak "
,"Fontana_z_weight "
,"Fontana_z683_low "
,"Fontana_z683_high "
,"Fontana_z954_low "
,"Fontana_z954_high "
,"Pforr_z_peak "
,"Pforr_z_weight "
,"Pforr_z683_low "
,"Pforr_z683_high "
,"Pforr_z954_low "
,"Pforr_z954_high "
,"Salvato_z_peak "
,"Salvato_z_weight "
,"Salvato_z683_low "
,"Salvato_z683_high "
,"Salvato_z954_low "
,"Salvato_z954_high "
,"Wiklind_z_peak "
,"Wiklind_z_weight "
,"Wiklind_z683_low "
,"Wiklind_z683_high "
,"Wiklind_z954_low "
,"Wiklind_z954_high "
,"Wuyts_z_peak "
,"Wuyts_z_weight "
,"Wuyts_z683_low "
,"Wuyts_z683_high "
,"Wuyts_z954_low "
,"Wuyts_z954_high" ]

##### Lo mostramos

In [10]:
phot_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41457 entries, 0 to 41456
Data columns (total 57 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0    file                   41457 non-null  object 
 1   ID                      41457 non-null  int64  
 2   RA                      41457 non-null  float64
 3   DEC                     41457 non-null  float64
 4   z_best                  41457 non-null  float64
 5   z_best_type             41457 non-null  object 
 6   z_spec                  41457 non-null  float64
 7   z_spec_ref              41457 non-null  object 
 8   z_grism                 41457 non-null  float64
 9   mFDa4_z_peak            41457 non-null  float64
 10  mFDa4_z_weight          41457 non-null  float64
 11  mFDa4_z683_low          41457 non-null  float64
 12  mFDa4_z683_high         41457 non-null  float64
 13  mFDa4_z954_low          41457 non-null  float64
 14  mFDa4_z954_high         41457 non-null

## Filtramos los datos

En esta parte basicamente, vamos a seleccionar el redshift que queremos en nuestro caso el del cosmic noon, ver que todos los valores son únicos y eliminar algunos datos que puedan considerarse como outliers, por ejemplo, los agujeros negros que son un pequeño número en comparación con las que no.

In [11]:
cosmic_noon_phot = phot_data[(1.5<phot_data["z_best "])& (phot_data["z_best "]<2.5)]
cosmic_noon_phot

file      ID          RA        DEC  \
0      ALL_OPTIMIZED_PDFS_EGS_ID00001.pzd      1  215.301103  53.051246   
6      ALL_OPTIMIZED_PDFS_EGS_ID00007.pzd      7  215.301035  53.051430   
13     ALL_OPTIMIZED_PDFS_EGS_ID00014.pzd     14  215.299993  53.051383   
15     ALL_OPTIMIZED_PDFS_EGS_ID00016.pzd     16  215.299495  53.051344   
17     ALL_OPTIMIZED_PDFS_EGS_ID00018.pzd     18  215.293744  53.047429   
...                                   ...    ...         ...        ...   
41449  ALL_OPTIMIZED_PDFS_EGS_ID41450.pzd  41450  214.580315  52.721460   
41452  ALL_OPTIMIZED_PDFS_EGS_ID41453.pzd  41453  214.586356  52.724939   
41454  ALL_OPTIMIZED_PDFS_EGS_ID41455.pzd  41455  214.719359  52.819680   
41455  ALL_OPTIMIZED_PDFS_EGS_ID41456.pzd  41456  214.887465  52.928631   
41456  ALL_OPTIMIZED_PDFS_EGS_ID41457.pzd  41457  214.668002  52.769395   

       z_best  z_best_type   z_spec  z_spec_ref   z_grism   mFDa4_z_peak   \
0        1.947            p     -1.0          --      -1.0           1.95   
6        1.567            p     -1.0          --      -1.0           1.55   
13       1.589            p     -1.0          --      -1.0           1.53   
15       1.879            p     -1.0          --      -1.0           1.60   
17       1.788            p     -1.0          --      -1.0           1.99   
...        ...          ...      ...         ...       ...            ...   
41449    1.702            p     -1.0          --      -1.0           1.50   
41452    1.912            p     -1.0          --      -1.0           2.00   
41454    2.078            p     -1.0          --      -1.0           1.74   
41455    1.590            p     -1.0          --      -1.0           1.45   
41456    1.911            p     -1.0          --      -1.0           2.55   

       ...  Wiklind_z683_low   Wiklind_z683_high   Wiklind_z954_low   \
0      ...              1.683               2.204              1.536   
6      ...              1.645               6.789              1.453   
13     ...              0.443               2.734              0.036   
15     ...              1.235               2.476              0.682   
17     ...              1.805               2.188              1.633   
...    ...                ...                 ...                ...   
41449  ...              0.226               2.190              0.032   
41452  ...              0.364               2.590              0.075   
41454  ...              0.847               5.271              0.129   
41455  ...              0.375               1.990              0.045   
41456  ...              0.499               2.790              0.085   

       Wiklind_z954_high   Wuyts_z_peak   Wuyts_z_weight   Wuyts_z683_low   \
0                   2.575           1.94            1.934            1.879   
6                   9.641           1.50            1.522            1.432   
13                  5.188           1.53            1.589            0.951   
15                  2.945           1.19            1.435            1.006   
17                  2.375           2.58            2.624            2.574   
...                   ...            ...              ...              ...   
41449               2.671           1.50            1.702            1.414   
41452               2.862           2.00            1.912            1.528   
41454               8.512           1.73            1.854            0.933   
41455               2.468           1.45            1.590            1.266   
41456               3.409           2.07            2.002            1.433   

       Wuyts_z683_high   Wuyts_z954_low   Wuyts_z954_high  
0                 1.968            1.492            2.002  
6                 1.618            1.351            1.774  
13                2.206            0.546            2.702  
15                1.890            0.837            2.139  
17                2.687            2.554            2.767  
...                 ...              ...        

In [12]:
import json
def filter_redsift(phot_data,mass_data,phys_prop_data,z =[(0.0,0.4),(0.8,1.2),(1.5,2.5)]):
    db = {} # Creamos un diccionario vacío
    # Usamos los valores de z para cada rango de z y así ver cuantos datos hay
    for z_min, z_max in z:
        a = phot_data[(z_min<=phot_data["z_best "])& (phot_data["z_best "]<=z_max)]
        b = mass_data[(z_min<=phot_data["z_best "])& (phot_data["z_best "]<=z_max)]
        c = phys_prop_data[(z_min<=phot_data["z_best "])& (phot_data["z_best "]<=z_max)]
    
        db['Mass z='+str(z_min)+'-'+str(z_max)]= f"Columns:{a.shape[1]} Rows:{a.shape[0]}" 
        db['Photo z='+str(z_min)+'-'+str(z_max)]= f"Columns:{b.shape[1]} Rows:{b.shape[0]}"
        db['Phys_prop z='+str(z_min)+'-'+str(z_max)]= f"Columns:{c.shape[1]} Rows:{c.shape[0]}"
        
        
        
        
        
    
    
    return db

db = filter_redsift(phot_data,mass_data,phys_prop_data)
print(json.dumps(db,indent=2))

{
  "Mass z=0.0-0.4": "Columns:57 Rows:4517",
  "Photo z=0.0-0.4": "Columns:39 Rows:4517",
  "Phys_prop z=0.0-0.4": "Columns:123 Rows:4517",
  "Mass z=0.8-1.2": "Columns:57 Rows:4683",
  "Photo z=0.8-1.2": "Columns:39 Rows:4683",
  "Phys_prop z=0.8-1.2": "Columns:123 Rows:4683",
  "Mass z=1.5-2.5": "Columns:57 Rows:15162",
  "Photo z=1.5-2.5": "Columns:39 Rows:15162",
  "Phys_prop z=1.5-2.5": "Columns:123 Rows:15162"
}


In [13]:
def filter_redsift(z_min,z_max,phot_data,mass_data,phys_prop_data):
    """ Esta función filtra los datos para un rango de valores de z. Teniendo en cuenta que los valores de z son diferentes 
    para datasets de Photometry y Mass y que Physical Properties no tiene este dato, por lo que el filtro consiste en que esté en unos de los otros dos datasets. """
    a = phot_data[(z_min<=phot_data["z_best "])& (phot_data["z_best "]<=z_max)]
    b = mass_data[(z_min<=mass_data["zbestM "])& (mass_data["zbestM "]<=z_max)]
    c = phys_prop_data[phys_prop_data["ID"].isin(pd.concat([a["ID"], b["ID"]]).unique())]
    return a, b, c

cosmic_noon_phot, cosmic_noon_mass, cosmic_noon_phys_prop = filter_redsift(1.5,2.5,phot_data,mass_data,phys_prop_data)

In [14]:
cosmic_noon_phys_prop.head()

ID  age_2a_tau (dex(yr))  tau_2a_tau (Gyr)  Av_2a_tau (mag)  \
0    1                   7.6            12.589              3.2   
6    7                   9.3             0.316              0.0   
13  14                   9.9             1.995              1.1   
15  16                   7.9             0.794              1.8   
17  18                   7.6             0.316              2.0   

    SFR_2a_tau (solMass/yr)  chi2_2a_tau   age_4b (dex(yr))  EBV_4b (mag)  \
0                 102.30000         2.830              8.31          0.42   
6                   0.03802         1.610              9.48          0.06   
13                  0.01000         0.514              9.90          0.06   
15                 21.38000         0.981              7.30          0.42   
17                112.20000        32.700              7.00          0.39   

    age_6a_tau (dex(yr))  tau_6a_tau (Gyr)  ...  age_14a_deltau (dex(yr))  \
0                   9.06               0.6  ...                     8.988   
6                   9.36               0.1  ...                     9.005   
13                  7.10               5.0  ...                     8.709   
15                  9.34              15.0  ...                     8.732   
17                  7.50               2.0  ...                     8.130   

    tau_14a_deltau (Gyr)  EBV_14a_deltau (mag)  SFR_14a_deltau (solMass/yr)  \
0                  0.147                 0.196                       1.2430   
6                  0.143                 0.029                       0.2839   
13                 0.366                 0.575                       0.3728   
15                 0.662                 0.347                      10.1800   
17                 0.179                 0.478                      78.7400   

    q_14a_deltau   age_14a_tau (dex(yr))  tau_14a_tau (Gyr)  \
0               2                  8.888              0.146   
6               2                  9.023              0.160   
13              1                  8.741              0.663   
15              1                  8.571              0.773   
17              1                  8.054              0.721   

    EBV_14a_tau (mag)  SFR_14a_tau (solMass/yr)  q_14a_tau   
0               0.217                   0.67120           2  
6               0.027                   0.06461           2  
13              0.552                   0.22920           1  
15              0.332                   7.86300           1  
17              0.468                  64.22000           1  

[5 rows x 123 columns]

Ahora vamos a juntarlos todos en un solo dataframe, dado que los datos de fotometría y masa son diferentes entre si, queremos quedarnos con las galaxias que estén dentro de los rangos de redshift, y que se encuentren en ambos datasets. Cuando comparemos medidas de fotometría y masa habrá que usar el redshift correspondiente de cada dataset.

En segundo lugar, nos quedamos con las propiedades físicas de de las galaxias que están ahí dentro de los rangos de redshift, y que se encuentren en ambos datasets. Esto nos permitirá comparar las propiedades físicas de las galaxias en ambos datasets.

#### A todo esto, se ha asumido que el campo de ID es el mismo y unequivoco en todos los datasets.



In [15]:
merged1 = pd.merge(cosmic_noon_phot,cosmic_noon_mass, left_on = 'ID', right_on = 'ID', how = 'inner')

merged2 = pd.merge(cosmic_noon_phys_prop,merged1, left_on = 'ID', right_on = 'ID', how = 'inner')

merged2.head(10)

ID  age_2a_tau (dex(yr))  tau_2a_tau (Gyr)  Av_2a_tau (mag)  \
0   1                   7.6            12.589              3.2   
1   7                   9.3             0.316              0.0   
2  16                   7.9             0.794              1.8   
3  18                   7.6             0.316              2.0   
4  19                   8.9             0.316              1.4   
5  20                   7.6            12.589              0.6   
6  21                   8.9             0.316              0.4   
7  23                   8.6             0.316              0.8   
8  40                   8.9             0.316              0.6   
9  42                   8.5             0.501              0.6   

   SFR_2a_tau (solMass/yr)  chi2_2a_tau   age_4b (dex(yr))  EBV_4b (mag)  \
0                102.30000         2.830              8.31          0.42   
1                  0.03802         1.610              9.48          0.06   
2                 21.38000         0.981              7.30          0.42   
3                112.20000        32.700              7.00          0.39   
4                 11.48000         3.000              8.61          0.30   
5                  3.80200         7.020              7.00          0.18   
6                  0.54950         1.720              7.00          0.36   
7                  1.69800         0.530              8.01          0.39   
8                  2.23900         0.823              8.46          0.24   
9                  5.01200         1.170              7.00          0.27   

   age_6a_tau (dex(yr))  tau_6a_tau (Gyr)  ...  M_10c (dex(solMass))   \
0                  9.06               0.6  ...                  10.18   
1                  9.36               0.1  ...                  10.15   
2                  9.34              15.0  ...                   9.37   
3                  7.50               2.0  ...                  10.09   
4                  8.90               0.3  ...                  10.21   
5                  8.50               0.3  ...                   8.51   
6                  8.80               0.3  ...                   8.81   
7                  8.90               0.6  ...                   8.80   
8                  9.28               1.0  ...                   9.47   
9                  8.50              15.0  ...                   8.89   

   M_14a_lin (dex(solMass))   M_14a_deltau (dex(solMass))   \
0                      10.30                         10.10   
1                       9.69                          9.62   
2                       9.39                          9.41   
3                       9.75                          9.74   
4                      10.29                         10.43   
5                       8.32                          8.32   
6                       8.48                          8.49   
7                       8.75                          8.79   
8                       9.48                          9.59   
9                       8.50                          8.51   

   M_14a_tau (dex(solMass))   M_14a_inctau (dex(solMass))   \
0                      10.09                         10.28   
1                       9.66                          9.74   
2                       9.42                          9.40   
3                       9.79                          9.77   
4                      10.36                         10.35   
5                       8.43                          8.40   
6                       8.59                          8.54   
7                       8.84                          8.79   
8                       9.61                          9.52   
9                       8.63                          8.58   

   M_14a (dex(solMass))   M_neb_med_lin (solMass)   s_neb_med_lin (solMass)   \
0                  10.10                     10.10                    0.2120   
1                   9.62                      9.68                    0.1230   
2                   9.44     

In [16]:
# Ahora eliminimaos aquellas galaxias que tienen un AGN, ya que son pocas y no podemos comparar su comportamiento con el de otras que no tienen AGN.
value_to_drop = 1
column_name = 'AGNflag '


df_cleaned = merged2.drop(merged2[merged2[column_name] == value_to_drop].index)

df_cleaned.head()


ID  age_2a_tau (dex(yr))  tau_2a_tau (Gyr)  Av_2a_tau (mag)  \
0   1                   7.6            12.589              3.2   
1   7                   9.3             0.316              0.0   
2  16                   7.9             0.794              1.8   
3  18                   7.6             0.316              2.0   
4  19                   8.9             0.316              1.4   

   SFR_2a_tau (solMass/yr)  chi2_2a_tau   age_4b (dex(yr))  EBV_4b (mag)  \
0                102.30000         2.830              8.31          0.42   
1                  0.03802         1.610              9.48          0.06   
2                 21.38000         0.981              7.30          0.42   
3                112.20000        32.700              7.00          0.39   
4                 11.48000         3.000              8.61          0.30   

   age_6a_tau (dex(yr))  tau_6a_tau (Gyr)  ...  M_10c (dex(solMass))   \
0                  9.06               0.6  ...                  10.18   
1                  9.36               0.1  ...                  10.15   
2                  9.34              15.0  ...                   9.37   
3                  7.50               2.0  ...                  10.09   
4                  8.90               0.3  ...                  10.21   

   M_14a_lin (dex(solMass))   M_14a_deltau (dex(solMass))   \
0                      10.30                         10.10   
1                       9.69                          9.62   
2                       9.39                          9.41   
3                       9.75                          9.74   
4                      10.29                         10.43   

   M_14a_tau (dex(solMass))   M_14a_inctau (dex(solMass))   \
0                      10.09                         10.28   
1                       9.66                          9.74   
2                       9.42                          9.40   
3                       9.79                          9.77   
4                      10.36                         10.35   

   M_14a (dex(solMass))   M_neb_med_lin (solMass)   s_neb_med_lin (solMass)   \
0                  10.10                     10.10                     0.212   
1                   9.62                      9.68                     0.123   
2                   9.44                      9.74                     0.168   
3                   9.77                      9.77                     0.240   
4                  10.36                     10.40                     0.069   

   M_med_lin (solMass)   s_med_lin (solMass)  
0                 10.00               0.1080  
1                  9.69               0.0694  
2                  9.43               0.1090  
3                  9.45               0.3840  
4                 10.40               0.3580  

[5 rows x 217 columns]

Ahora solo nos quedan 217 columnas, por lo que habrá que ver cuales se repiten y cuales tienen un sentido físico propio y útil para el análisis.

In [17]:
df_cleaned[['z_best ', 'zbestM ']].head()

z_best   zbestM 
0    1.947    1.700
1    1.567    1.611
2    1.879    1.673
3    1.788    1.987
4    1.750    1.762

El mayor problema ahora es que tenemos 2 valores de mejor redshift, uno para los datos fotométricos y otro para los datoss de masa

In [18]:
df_cleaned[['RA', 'RAdeg (deg) ']].head()

RA  RAdeg (deg) 
0  215.301103    215.301100
1  215.301035    215.301033
2  215.299495    215.299493
3  215.293744    215.293741
4  215.240631    215.240622

In [19]:
df_cleaned[['DEC', 'DECdeg (deg) ']].head()

DEC  DECdeg (deg) 
0  53.051246      53.051206
1  53.051430      53.051390
2  53.051344      53.051304
3  53.047429      53.047388
4  53.009947      53.009898

In [ ]:
from astropy.coordinates import SkyCoord
import astropy.units as u
import plotly.graph_objects as go
import matplotlib.pyplot as plt

def eq2gal(ra, dec):
    
    '''
    Transforms equatorial coordinates to galactic ones.
    Then prepares them for matplotlib aitoff projection. 
    '''
    
    eq = SkyCoord(ra, dec, unit=u.deg)
    gal = eq.galactic

    # Minus appears because of “mapping from the inside” issue
    l_gal, b_gal = -gal.l.wrap_at('180d').radian, gal.b.radian
    
    return l_gal, b_gal


ra, dec = eq2gal(df_cleaned['RA'], df_cleaned['DEC'])
fig = go.Figure()

fig.add_trace(go.Scattergeo(
    mode="markers",
    lon = ra,
    lat = dec,
    marker = {'size': 8,
              'color':'blue',
             })
             )
fig.update_layout(geo=dict(
                    showland=False,
                    showcountries=False,
                    showocean=False,
                    countrywidth=0.5,
                    #coastlinecolor='rgb(255, 255, 255)',
                    landcolor='rgb(255, 255, 255)',
                    lakecolor='rgb(255, 255, 255)',
                    oceancolor='rgb(255, 255, 255)',
                    projection=dict(
                        type='aitoff',
                    ),
                    lonaxis=dict(
                        showgrid=True,
                        dtick=15,
                        gridcolor='rgb(102, 102, 102)',
                        gridwidth=0.5
                    ),
                    lataxis=dict(
                        showgrid=True,
                        dtick=15,
                        gridcolor='rgb(102, 102, 102)',
                        gridwidth=0.5
                    )
                ),autosize=False,
    width=1000,
    height=1000,
)
fig.show()


Estan las 11.000 galaxias en 1 punto enano en la tierra, bastante impresionante

Podemos eliminar una de estas columnas ya que son iguales, luego habrá que ver que más columnas podemos quitar como los z que no son zbest

Lista de columnas a eliminar:
    - DEC
    - RA
    - AGNflag
    - file

Ahora vamos a estudiar a un solo autor, luego si podemos estudiamos al resto de autores.

In [26]:
fontana = df_cleaned.filter(regex = r'6a|^ID|^z_best|^zbestM|^RAdeg|^DECdeg$', axis=1)
fontana.head()

ID  age_6a_tau (dex(yr))  tau_6a_tau (Gyr)  EBV_6a_tau (mag)  \
0   1                  9.06               0.6              0.45   
1   7                  9.36               0.1              0.00   
2  16                  9.34              15.0              0.20   
3  18                  7.50               2.0              0.45   
4  19                  8.90               0.3              0.35   

   SFR_6a_tau (solMass/yr)  met_6a_tau (solMet)  extlw_6a_tau   chi2_6a_tau   \
0                    5.656                  2.5              2         8.260   
1                    0.000                  0.2              2         5.820   
2                    2.839                  2.5              2         0.858   
3                   80.020                  1.0              1         6.690   
4                   10.630                  1.0              1         1.150   

   L1400_6a_tau (erg/s/Hz)  L2700_6a_tau (erg/s/Hz)  ...  \
0             1.450000e+26             5.100000e+27  ...   
1             2.760000e+25             3.070000e+27  ...   
2             1.890000e+27             1.020000e+28  ...   
3             7.890000e+27             2.950000e+28  ...   
4             4.270000e+27             1.670000e+28  ...   

   RMag_6a_invtau (mag)  IMag_6a_invtau (mag)  JMag_6a_invtau (mag)  \
0                -20.73                -21.17                -21.92   
1                -20.26                -20.60                -20.96   
2                -20.11                -20.42                -21.02   
3                -20.98                -21.32                -22.00   
4                -21.80                -22.19                -22.80   

   KMag_6a_invtau (mag)  z_best   z_best_type   RAdeg (deg)   zbestM   \
0                -22.00    1.947             p    215.301100    1.700   
1                -20.59    1.567             p    215.301033    1.611   
2                -21.12    1.879             p    215.299493    1.673   
3                -22.16    1.788             p    215.293741    1.987   
4                -22.92    1.750             p    215.240622    1.762   

   M_6a_tau^NEB (dex(solMass))   M_6a_tau (dex(solMass))   
0                         10.14                     10.08  
1                          9.77                      9.77  
2                          9.74                      9.61  
3                          9.94                      9.35  
4                         10.46                     10.45  

[5 rows x 71 columns]

# Creamos unos repots en HTML para explorar los datos

#import pandas as pd

from tqdm._tqdm import tqdm 
from ydata_profiling import ProfileReport
from IPython.display import display

# Create a DataFrame


# Generate the profile report
#profile1 = ProfileReport(cosmic_noon_phys_prop, title="Physical properties of the galaxies in COSMIC Noon")
#
#profile1.to_file("cosmic_noon_phys_prop_report.html")


# Embed HTML file using IFrame
file_path = 'cosmic_noon_mass_prop_report.html'  
from IPython.core.display import HTML


with open(file_path, 'r') as f:
    html_content = f.read()

display(HTML(html_content))


In [ ]:
#profile2 = ProfileReport(cosmic_noon_phot, title="Redshift of the galaxies in COSMIC Noon")
#profile2.to_file("cosmic_noon_phot_prop_report.html") 


In [ ]:
#profile3 = ProfileReport(cosmic_noon_mass, title="Mass of the galaxies in COSMIC Noon")
#profile3.to_file("cosmic_noon_mass_prop_report.html")

In [ ]:
profile4 = ProfileReport(cosmic_noon_mass, title="Mass of the galaxies in COSMIC Noon")
profile4.to_file("cosmic_noon_mass_prop_report.html")

NameError: name 'ProfileReport' is not defined

In [ ]:
cosmic_noon_phys_prop

#hue = cosmic_noon_mass[["",""]]
mas_cols=[i for i in mass_data.columns[21:35]]

g = sns.PairGrid(cosmic_noon_mass[mas_cols]) 
g.map_diag(sns.histplot)
g.map_offdiag(sns.scatterplot, hue=cosmic_noon_mass["AGNflag        "])
g.add_legend(title="", adjust_subtitles=True)

In [ ]:
def translate_column_name(column_name):
    # Dictionaries for mapping codes and SFH types
    pi_codes = {
        '2a': 'Barro',
        '6a': 'Fontana',
        '10a': 'Pforr',
        '11a': 'Salvato',
        '12a': 'Wiklind',
        '13a': 'Wuyts',
        '14a': 'BooMe Lee',
        '15a': 'Lee',
    }
    
    sfh_types = {
        'cons': 'constant',
        'tau': 'exponentially declining',
        'deltau': 'delayed-exponential',
        'lin': 'linearly increasing',
        'inctau': 'exponentially increasing',
    }
    
    # Split the input string to isolate the relevant part
    relevant_part = column_name.split(' ')[0]
    
    try:
        # Split the relevant part into components
        _, pi_code, sfh_code = relevant_part.split('_')
        
        # Translate using the dictionaries
        pi_translation = pi_codes.get(pi_code, 'Unknown PI code')
        sfh_translation = sfh_types.get(sfh_code, 'Unknown SFH type')
        
        # Create a dictionary for the output
        
        return {
            "PI": pi_translation,
            "SFH": sfh_translation
        }
    except ValueError:
        return {"error": "Invalid column name format"}

# Example usage
column_name = "M_2a_tau   (dex(solMass)))"
translated = translate_column_name(column_name)
print(translated)  # Output: {'PI': 'Barro', 'SFH': 'exponentially declining'}


In [ ]:
# Mostrar estadísticas descriptivas para las columnas 21 a 34 (Python usa índices de 0)
mass_data_subset = mass_data.iloc[:, 21:35]
filtered_data = mass_data_subset[(mass_data_subset > 0).all(axis=1)]
filtered_data[:5]


In [ ]:
for i in mass_data.columns[21:35]:
    g = sns.JointGrid(data=cosmic_noon_mass[cosmic_noon_mass[i]>0], x="zbest          ", y=i)
    trans = translate_column_name(i)
    title = f"Mass vs redshift from {trans['PI']} using {trans['SFH']}"
    g.plot(sns.regplot, sns.histplot)
    plt.suptitle(title, y = 1)


In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Crear el pairplot para todas las columnas seleccionadas
sns.pairplot(filtered_data)

# Título del gráfico
plt.suptitle('Comparación de columnas entre sí', y=1.02)

# Mostrar el gráfico
plt.show()



In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Calcula la matriz de correlación
correlation_matrix = filtered_data.corr()

# Grafica el mapa de calor
plt.figure(figsize=(10, 8))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', vmin=-1, vmax=1)
plt.title("Mapa de Calor de Correlación")
plt.show()


In [ ]:
# Selecciona las columnas con correlación alta
high_corr_columns = correlation_matrix.columns[(correlation_matrix > 0.9).any() | (correlation_matrix < -0.9).any()]
filtered_df = filtered_data[high_corr_columns]

# Grafica el pairplot solo para las columnas seleccionadas
sns.pairplot(filtered_df)
plt.suptitle("Pairplot Filtrado (Solo Correlaciones Significativas)", y=1.02)
plt.show()


# Dendograma de la masa de los diferentes autores

In [ ]:
from scipy.cluster.hierarchy import dendrogram, linkage

# Calcula las distancias y el linkage para el dendrograma
linked = linkage(filtered_data.T, method='ward', optimal_ordering=True)  # 'ward' es común, pero se pueden probar otros métodos

# Grafica el dendrograma
plt.figure(figsize=(10, 6))
dendrogram(linked, labels=filtered_data.columns, orientation='top', leaf_rotation=90)
plt.title("Dendrograma de Similitud entre la masa de los autores")
plt.show()


# PCA

In [ ]:
from sklearn.decomposition import PCA
import numpy as np

# Normaliza los datos (opcional pero recomendado para PCA)
df_normalized = (filtered_data - filtered_data.mean()) / filtered_data.std()

# Aplica PCA
pca = PCA(n_components=2)
pca_result = pca.fit_transform(df_normalized.T)  # Transpuesta para comparar columnas en lugar de filas

# Grafica los resultados de PCA
plt.figure(figsize=(8, 6))
plt.scatter(pca_result[:, 0], pca_result[:, 1])

# Añade etiquetas a cada punto
for i, column in enumerate(filtered_data.columns):
    plt.text(pca_result[i, 0], pca_result[i, 1], column)

plt.title("Análisis de Componentes Principales (PCA) de Columnas")
plt.xlabel("Componente Principal 1")
plt.ylabel("Componente Principal 2")
plt.show()


In [ ]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit

# Definir funciones para el ajuste de curvas
def linear(x, a, b):
    return a * x + b

def quadratic(x, a, b, c):
    return a * x**2 + b * x + c

def exponential(x, a, b):
    return a * np.exp(b * x)

# Seleccionar dos columnas para comparar (por ejemplo, 'col1' y 'col2')
for i in range(len(filtered_data.columns)-1):
    p = filtered_data.columns[i]
    s = filtered_data.columns[i+1]
    x = filtered_data[p]
    y = filtered_data[s]

    # Gráfico de dispersión
    plt.scatter(x, y, label="Datos")

    # Ajuste de curva lineal
    params, _ = curve_fit(linear, x, y)
    plt.plot(x, linear(x, *params), label="Lineal", color='red')

    # Ajuste de curva cuadrática
    params, _ = curve_fit(quadratic, x, y)
    plt.plot(x, quadratic(x, *params), label="Cuadrática", color='green')

    # Ajuste de curva exponencial
    params, _ = curve_fit(exponential, x, y)
    plt.plot(x, exponential(x, *params), label="Exponencial", color='blue')

    # Personalización del gráfico
    plt.legend()
    plt.xlabel(filtered_data.columns[i])
    plt.ylabel(filtered_data.columns[i+1])
    plt.title("Ajuste de Curvas: Lineal, Cuadrática y Exponencial")
    plt.show()


In [ ]:
for i in range(len(filtered_data.columns)-1):
    p = filtered_data.columns[i]
    s = filtered_data.columns[i+1]
    x = filtered_data[p]
    y = filtered_data[s]

# Correlación de Spearman
    spearman_corr = x.corr(y, method='spearman')
    print(f"Correlación de Spearman de {p} con {x}: {spearman_corr}")

    # Correlación de Kendall
    kendall_corr = x.corr(y, method='kendall')
    print(f"Correlación de Kendall de {p} con {x}: {kendall_corr}")


In [ ]:
from sklearn.metrics import r2_score

# Ajuste lineal
for i in range(len(filtered_data.columns)-1):
    p = filtered_data.columns[i]
    s = filtered_data.columns[i+1]
    x = filtered_data[p]
    y = filtered_data[s]
    params, _ = curve_fit(linear, x, y)
    y_pred_linear = linear(x, *params)
    r2_linear = r2_score(y, y_pred_linear)

    # Ajuste cuadrático
    params, _ = curve_fit(quadratic, x, y)
    y_pred_quadratic = quadratic(x, *params)
    r2_quadratic = r2_score(y, y_pred_quadratic)

    # Ajuste exponencial
    params, _ = curve_fit(exponential, x, y)
    y_pred_exponential = exponential(x, *params)
    r2_exponential = r2_score(y, y_pred_exponential)
    print(f"{p} vs {s}:")
    if r2_linear >0.9 or r2_quadratic >0.9 or r2_exponential >0.9:
        print("R² Lineal:", r2_linear)
        print("R² Cuadrática:", r2_quadratic)
        print("R² Exponencial:", r2_exponential)
    else:
        print("R² muy bajo")


In [ ]:

for i in range(len(filtered_data.columns)-1):
    p = filtered_data.columns[i]
    s = filtered_data.columns[i+1]
    
# Regresión lineal (order=1)
    sns.regplot(x=p, y=s, data=filtered_data, order=1, line_kws={"color": "red"}, label="Lineal")

    # Regresión cuadrática (order=2)
    sns.regplot(x=p, y=s, data=filtered_data, order=2, line_kws={"color": "green"}, label="Cuadrática")

    # Regresión cúbica (order=3)
    sns.regplot(x=p, y=s, data=filtered_data, order=3, line_kws={"color": "blue"}, label="Cúbica")

    plt.legend()
    plt.title("Ajuste de Regresiones de Diferentes Órdenes")
    plt.show()


In [ ]:
import matplotlib.pyplot as plt
import numpy as np

r = np.arange(0, 2, 0.01)
theta = 2 * np.pi * r

fig, ax = plt.subplots(subplot_kw={'projection': 'polar'})
ax.plot(theta, r)
ax.set_rmax(2)
ax.set_rticks([0.5, 1, 1.5, 2])  # Less radial ticks
ax.set_rlabel_position(-22.5)  # Move radial labels away from plotted line
ax.grid(True)

ax.set_title("A line plot on a polar axis", va='bottom')
plt.show()